### PR TITLE
feat(logging): RHICOMPL-1217 introduce audit logging 

### PR DESCRIPTION
**Pull request by vkrizan**
Introduction of a middleware and utility to add entries to an audit log.  By default it logs them to STDOUT, but eventually it can be tied into CloudWatch. Log entries are formatted into JSON, split by a new-line.

Originally posted at https://github.com/RedHatInsights/insights-api-common-rails/pull/217

TODO:
- [X] Capture Rails requests
- [x] Add tests
- [ ] Find a way to have a shortcut to the logger
- [x] Support within a Sidekiq

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [X] Input Validation
- [X] Output Encoding
- [X] Authentication and Password Management
- [X] Session Management
- [X] Access Control
- [X] Cryptographic Practices
- [X] Error Handling and Logging
- [X] Data Protection
- [X] Communication Security
- [X] System Configuration
- [X] Database Security
- [X] File Management
- [X] Memory Management
- [X] General Coding Practices

### CODE REVIEW

**Review by akofink — COMMENTED** [review]
Looking good! I've already reviewed this on https://github.com/RedHatInsights/insights-api-common-rails/pull/217. Thanks, @vkrizan!

File: lib/audit_log/audit_log/middleware.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+# Common space for Insights API stuff
+module Insights
+  module API
+    module Common
+      class AuditLog
+        # Request events listener, capturing for evidence
+        class Middleware
```
Comment (akofink) [inline lib/audit_log/audit_log/middleware.rb:9]: This represents the top level object, the middleware itself. If I understand correctly, it's intended to use this like:

```rb
config.middleware.use Insights::API::Common::AuditLog::Middleware
```

If that's the case, we need to ensure all the `require`s start from here so things like `AuditLog.logger` are already defined properly.

File: test/lib/audit_log/middleware_test.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+require 'audit_log/audit_log'
```
Comment (akofink) [inline test/lib/audit_log/middleware_test.rb:3]: I suppose this is the `require` intended to be used within `application.rb`?

**Review by akofink — COMMENTED** [review]
Looking great, @vkrizan! Just a couple small comments about naming, paths, etc., but I was able to enable the middleware without much fuss locally. Here's what a log look like:

```
{"@timestamp":"2021-02-09T23:05:15.780582Z","hostname":"dfa3fe94f424","pid":1,"thread_id":"2b0cf4a3925c","level":"warning","transaction_id":"57ba8762f3b44f8b08a3","message":"GET /api/compliance/v1/systems/435c3df2-4e81-4986-8462-8a9ba930c512 -> 404 Not Found","controller":"V1::SystemsController#show","remote_ip":"127.0.0.1"}
```

Are you intending to enable the logger in this PR? It'd certainly help to review if it was enabled. I'd love if we could get this going to `log/audit.log` in `Rails.env == 'development'` for debugging purposes (instead of mixed into STDOUT and `log/development.log` with all the rest).

File: lib/audit_log/audit_log.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+          def with_account(account_number)
+            original = Thread.current[:audit_account_number]
+            Thread.current[:audit_account_number] = account_number
+            return unless block_given?
+
+            begin
+              yield
+            ensure
+              Thread.current[:audit_account_number] = original
```
[the remark this answers is not in the document]
Comment (vkrizan) [reply lib/audit_log/audit_log.rb:34]: We can always go back and change it. :wink:

File: lib/audit_log/audit_log.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+require_relative 'audit_log/formatter'
+require_relative 'audit_log/middleware'
+
+# Common space for Insights API stuff
+module Insights
+  module API
```
[the remark this answers is not in the document]
Comment (vkrizan) [reply lib/audit_log/audit_log.rb:8]: Are you sure about that? There are Rails stuff named [`API` in `ActionController`](https://github.com/rails/rails/blob/v5.2.4.3/actionpack/lib/action_controller/api.rb#L89). If there are upper cased letters together then it concats them, not split them.

Test:
```
[1] pry(main)> 'API'.underscore
=> "api"
```
I can certainly move it to the appropriate path. I don't see other benefits other than the autoload, which can be neglected.

File: lib/audit_log/audit_log/middleware.rb (hunk 1 above)
Comment (vkrizan) [reply lib/audit_log/audit_log/middleware.rb:9]: Yes, the usage is correct.

The logger is setup with `Insights::API::Common::AuditLog.setup(Logger.new)`. Require has no effect on it, only the first use of the logger. The default one can be overridden by the mentioned setup.

**Comment by vkrizan** [comment]
It'd certainly help to review if it was enabled.

@akofink 

> Are you intending to enable the logger in this PR?

No. Please treat it as a library.

>  I'd love if we could get this going to log/audit.log in Rails.env == 'development' for debugging purposes (instead of mixed into STDOUT and log/development.log with all the rest).

Sure, you can use setup:
```
Insights::API::Common::AuditLog.setup(Logger.new('log/audit.log'))
```

File: test/lib/audit_log/middleware_test.rb (hunk 2 above)
Comment (vkrizan) [reply test/lib/audit_log/middleware_test.rb:3]: Yes.

File: lib/audit_log/audit_log.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (akofink) [reply lib/audit_log/audit_log.rb:8]: Yes, it's only for Rails autoloaded constants. It's not necessary to match this format if we manually `require` the correct file, which is how this code will ultimately be used. Also, +1 to matching the insights-api-common-rails. Just for the sake of argument :trollface: 

```rb
[21] pry(main)> 'api'.camelize
=> "Api"
[22] pry(main)> 'a_p_i'.camelize
=> "API"
```

This was important for our rails classes like `HostInventoryApi` which we expect to be autoloaded without a `require`.

**Review by akofink — APPROVED** [review]
Thanks, @vkrizan! This is an exciting addition to the application.

**Pull request merged**